### PR TITLE
[TBAA] Extend pointer TBAA to pointers of non-builtin types.

### PR DESCRIPTION
**Pull request by fhahn**
Extend the logic added in 123c036bd361d
(https://github.com/llvm/llvm-project/pull/76612) to support pointers to non-builtin types by using the mangled name of the canonical type.

### CODE REVIEW

**Comment by fhahn** [comment]
ping :)

File: clang/lib/CodeGen/CodeGenTBAA.cpp [hunk 1]
```diff
@@ -221,21 +221,27 @@ llvm::MDNode *CodeGenTBAA::getTypeInfoHelper(const Type *Ty) {
...
+
+    SmallString<256> TyName;
     if (isa<BuiltinType>(Ty)) {
       llvm::MDNode *ScalarMD = getTypeInfoHelper(Ty);
       StringRef Name =
           cast<llvm::MDString>(
               ScalarMD->getOperand(CodeGenOpts.NewStructPathTBAA ? 2 : 0))
               ->getString();
+      TyName = Name;
+    } else if (!isa<VariableArrayType>(Ty)) {
```
Comment (rjmccall) [inline clang/lib/CodeGen/CodeGenTBAA.cpp:233]: I feel like we should just ignore all array structure here, whether it's variable-length or fixed-length.  For one, partially honoring array structures probably violates the VLA compatibility rules — IIUC, the C standard requires l-values of type `T (*)[n]` to be able to alias objects of type `T (*)[10]` if `n == 10` dynamically, but with your patch, we'd now create a node for that pointer-to-constant-array type with no relation to the node we'd use for the pointer-to-VLA type.  But also, I feel like trying to honor the array structure is just pointlessly strict; we can treat a `T (*)[10]` object like a `T *` object, it's not really going to hurt anything.

You should look through the array structure before you do the builtin/non-builtin breakdown.  There's a `getBaseElementType` operation you can use.

**Review by rjmccall — CHANGES_REQUESTED** [review]

File: clang/lib/CodeGen/CodeGenTBAA.h [hunk 2]
```diff
@@ -119,6 +120,7 @@ class CodeGenTBAA {
   llvm::Module &Module;
   const CodeGenOptions &CodeGenOpts;
   const LangOptions &Features;
+  MangleContext *MangleCtx;
```
Comment (rjmccall) [inline clang/lib/CodeGen/CodeGenTBAA.h:123]: This should either be a unique_ptr or you need to delete it in the destructor.

**Review by rjmccall — COMMENTED** [review]
Thanks, LGTM other than the memory leak.

File: clang/lib/CodeGen/CodeGenTBAA.h (hunk 2 above)
Comment (fhahn) [reply clang/lib/CodeGen/CodeGenTBAA.h]: Thanks that should be fixed now!

**Review by rjmccall — APPROVED** [review]
LGTM

**Pull request merged**